server/src/com/thoughtworks/go/config/update/UpdateEnvironmentCommand.java

```diff
@@ -1,33 +1,83 @@
+import java.util.List;
 
-    private final String environmentToUpdate;
+import static com.thoughtworks.go.config.BasicCruiseConfig.copyErrors;
+
+public class UpdateEnvironmentCommand implements EntityConfigUpdateCommand<EnvironmentConfig> {
+
+
+    private GoConfigService goConfigService;
+    private final EnvironmentConfig oldEnvironmentConfig;
     private final EnvironmentConfig newEnvironmentConfig;
-    private Username username;
+    private final Username username;
+    private final HttpLocalizedOperationResult result;
+    private final List<GoConfigValidator> VALIDATORS = Arrays.asList(
+            new EnvironmentAgentValidator(),
+            new EnvironmentPipelineValidator()
+    );
 
-    public UpdateEnvironmentCommand(String environmentToUpdate, EnvironmentConfig newEnvironmentConfig, Username username) {
-        this.environmentToUpdate = environmentToUpdate;
+    public UpdateEnvironmentCommand(GoConfigService goConfigService, EnvironmentConfig oldEnvironmentConfig, EnvironmentConfig newEnvironmentConfig, Username username, HttpLocalizedOperationResult result) {
+        this.goConfigService = goConfigService;
+        this.oldEnvironmentConfig = oldEnvironmentConfig;
         this.newEnvironmentConfig = newEnvironmentConfig;
         this.username = username;
+        this.result = result;
     }
 
     @Override
-    public CruiseConfig update(CruiseConfig cruiseConfig) throws Exception {
-        EnvironmentsConfig environments = cruiseConfig.getEnvironments();
-        EnvironmentConfig oldConfig = environments.named(new CaseInsensitiveString(environmentToUpdate));
-        int index = environments.indexOf(oldConfig);
+    public void update(CruiseConfig preprocessedConfig) throws Exception {
+        EnvironmentsConfig environments = preprocessedConfig.getEnvironments();
+        int index = environments.indexOf(oldEnvironmentConfig);
         environments.remove(index);
         environments.add(index, newEnvironmentConfig);
-        return cruiseConfig;
     }
 
     @Override
-    public ConfigModifyingUser user() {
-        return new ConfigModifyingUser(username);
+    public boolean isValid(CruiseConfig preprocessedConfig) {
+        for (GoConfigValidator validator : VALIDATORS) {
+            try{
+                validator.validate(preprocessedConfig);
+            } catch (Exception e) {
+                Localizable.CurryableLocalizable actionFailed = LocalizedMessage.string("ENV_UPDATE_FAILED", oldEnvironmentConfig.name());
+                result.badRequest(actionFailed.addParam(e.getMessage()));
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public void clearErrors() {
+        BasicCruiseConfig.clearErrors(oldEnvironmentConfig);
+    }
+
+    @Override
+    public EnvironmentConfig getPreprocessedEntityConfig() {
+        return newEnvironmentConfig;
+    }
+
+    @Override
+    public boolean canContinue(CruiseConfig cruiseConfig) {
```

Comment:
Just adding this so we won't forget. :) After #2416 is merged, we can make the change to check for a stale request as well.